### DISCUSSION TITLE
after a deploy a change in the template is not applied

### DISCUSSION THREAD

**Discussion by remdan** [sulu/sulu#7555] [Q&A] [answered]
# Actual Behavior

after a deploy a change in the template is not applied, something is cached here. in the backend but also the preview and frontend.

# Expected Behavior

the current template should always be displayed in the admin.

# Steps to Reproduce

I deploy with deployer.org.
the new version is on the server. I can, for example, display new assets from the public-folder (with the direct url). changes made while editing in admin are applied and publishing is also possible.

**Comment by alexander-schranz** [comment]
Hello @remdan,

can you show your deployer config? And be more specific what you mean with template? Do you mean twig files or Sulu xml templates?

Did you try manually clear the caches on the server to find out which cache isn't cleared in your deployment process?

**Reply by remdan** [reply]
both,  twig files and Sulu xml templates...

```
<?php

namespace Deployer;

require 'recipe/sulu.php';

host('example')
    ->set('hostname', 'example.net')
    ->set('remote_user', 'example')
    ->set('deploy_path', '/home/example/www')
    ->set('branch', function () {
        return input()->getOption('branch') ?: 'main';
    })
    ->set('writable_mode', 'chmod')
    ->set('http_user', 'example')
    ->set('bin/git', 'git')
    ->set('bin/php', 'php')
    ->set('bin/composer', 'composer');

set('repository', 'git@gitlab.com:remdan/example.git');
set('composer_options', '--verbose --prefer-dist --no-progress --no-interaction --no-dev --optimize-autoloader --apcu-autoloader');

task('frontend_build', function () {
    runLocally('yarn encore production');
});

task('frontend_build:upload', function () {
    upload(__DIR__ . "/public/build/website", '{{release_path}}/public/build');
});

task('deploy:cache:clear_force', function () {
    run('rm -rf  {{release_path}}/var/cache/*');
});

before('deploy:vendors', 'frontend_build');
after('frontend_build', 'frontend_build:upload');
after('deploy:symlink', 'deploy:cache:clear_force');


// if deploy fails automatically unlock.
after('deploy:failed', 'deploy:unlock');
```

**Reply by remdan** [reply]
config/templates/pages/homepage.xml
templates/pages/homepage.html.twig

**Reply by alexander-schranz** [reply]
After deployment if you restart your webserver / php fpm is the new data then shown? Then its just a misconfigured VHost / Opcache configuration.

PS: `deploy:cache:clear_force` should not be called. The Symfony script already does run cache:clear. Make the `var/cache` empty without building it up before `deploy:symlink` can make your application crash in race conditions on high traffic servers. That is why deployer is important to be used and cache:clear with warmup is in deployer scripts done before symlinking and not after.

Can you show your Webserver and Opcache configuration? (PS: the Opcache from webserver (e.g. php-fpm php.ini) not the CLI).

**Reply by remdan** [reply]
```
opcache.blacklist_filename => no value => no value
opcache.consistency_checks => 0 => 0
opcache.dups_fix => Off => Off
opcache.enable => On => On
opcache.enable_cli => On => On
opcache.enable_file_override => Off => Off
opcache.error_log => no value => no value
opcache.file_cache => no value => no value
opcache.file_cache_consistency_checks => On => On
opcache.file_cache_only => Off => Off
opcache.file_update_protection => 2 => 2
opcache.force_restart_timeout => 180 => 180
opcache.huge_code_pages => Off => Off
opcache.interned_strings_buffer => 8 => 8
opcache.jit => tracing => tracing
opcache.jit_bisect_limit => 0 => 0
opcache.jit_blacklist_root_trace => 16 => 16
opcache.jit_blacklist_side_trace => 8 => 8
opcache.jit_buffer_size => 0 => 0
opcache.jit_debug => 0 => 0
opcache.jit_hot_func => 127 => 127
opcache.jit_hot_loop => 64 => 64
opcache.jit_hot_return => 8 => 8
opcache.jit_hot_side_exit => 8 => 8
opcache.jit_max_exit_counters => 8192 => 8192
opcache.jit_max_loop_unrolls => 8 => 8
opcache.jit_max_polymorphic_calls => 2 => 2
opcache.jit_max_recursive_calls => 2 => 2
opcache.jit_max_recursive_returns => 2 => 2
opcache.jit_max_root_traces => 1024 => 1024
opcache.jit_max_side_traces => 128 => 128
opcache.jit_prof_threshold => 0.005 => 0.005
opcache.lockfile_path => /tmp => /tmp
opcache.log_verbosity_level => 1 => 1
opcache.max_accelerated_files => 10000 => 10000
opcache.max_file_size => 0 => 0
opcache.max_wasted_percentage => 5 => 5
opcache.memory_consumption => 128 => 128
opcache.opt_debug_level => 0 => 0
opcache.optimization_level => 0x7FFEBFFF => 0x7FFEBFFF
opcache.preferred_memory_model => no value => no value
opcache.preload => no value => no value
opcache.preload_user => no value => no value
opcache.protect_memory => Off => Off
opcache.record_warnings => Off => Off
opcache.restrict_api => no value => no value
opcache.revalidate_freq => 2 => 2
opcache.revalidate_path => Off => Off
opcache.save_comments => On => On
opcache.use_cwd => On => On
opcache.validate_permission => Off => Off
opcache.validate_root => Off => Off
opcache.validate_timestamps => On => On
```

**Reply by alexander-schranz** [reply]
Looks good but is that really the one used by the webserver? Looks like CLI output you posted and on some installations like example Ubuntu/Debian there are 2 different php ini one for webserver one for cli.

If you manually run 
bin/console cache:clear
and
bin/websiteconsole cache:clear

your changes appear?

PS: do you use any external caching mechanism? Redis? And your browser cache isnt the one returning old website data?  

You should first find out which cache your deploy process doesnt clear. By try run all different Cache clears and service restarts manually until the Changes appear to know what causing the issue.

**Reply by remdan** [reply]
@alexander-schranz: you are right, after deployment and restart php the new data is shown!

<img width="906" alt="Bildschirmfoto 2024-08-03 um 10 36 11" src="https://github.com/user-attachments/assets/2ba82eed-0def-4900-ac71-38db0814279d">

**Reply by remdan** [reply]
[
<img width="871" alt="Bildschirmfoto 2024-08-03 um 10 55 34" src="https://github.com/user-attachments/assets/ce948fcc-c1ed-425f-be89-8f13f26a9184">
](url)

**Reply by alexander-schranz** [reply]
Its strange if you have opcache.validate_timestamp = On but it seems not detect that change. But at the end I think your Vhost is wrong check the Sulu VHost https://docs.sulu.io/en/latest/cookbook/web-server/nginx.html example or and deployer documentation: https://deployer.org/docs/7.x/avoid-php-fpm-reloading#fix-for-nginx

**Comment by NeuralClone** [comment, marked as answer]
I had a similar issue when using Deployer. My production server has to run Apache and it doesn't look for the new symlink Deployer creates without reloading PHP-FPM or clearing the OPCache. The old release path continues to be used.

I ended up using [CacheTool](https://github.com/gordalina/cachetool) and its associated [Deployer recipe](https://deployer.org/docs/7.x/contrib/cachetool) to clear the OPCache.

Make sure to run the `cachetool:clear:opcache` task after `deploy:symlink`.

Setting `opcache.revalidate_path = 1` in `php.ini` might work too. But I couldn't get it to work in my case.

**Comment by remdan** [comment]
thanks @NeuralClone your suggestion has helped.

opcache.revalidate_path = 1 in php.ini works for me
